maketeams3.py

```diff
@@ -258,60 +313,219 @@ def testSwap(teama, playera, teamb, playerb, board):
+        return n
+
+    count_friends_on_team = partial(count_on_team, 'friends')
+    count_avoids_on_team = partial(count_on_team, 'avoid')
+
+    pa, pb = swap
+    if count_friends_on_team(pa, pa.team) != count_friends_on_team(pa, pb.team)\
+            or count_friends_on_team(pb, pb.team) != count_friends_on_team(pb, pa.team)\
+            or count_avoids_on_team(pa, pa.team) != count_avoids_on_team(pa, pb.team)\
+            or count_avoids_on_team(pb, pb.team) != count_avoids_on_team(pb, pa.team):
+        return False
+
+    return True
+
+
+def get_swaps(teams):
+    num_boards = len(teams[0].boards)
+    boards = [[team.boards[i] for team in teams] for i in range(num_boards)]
+    swaps = [[swap for swap in combinations(board, 2) if is_neutral_swap(swap)] for board in boards]
+    return [swap for board_swaps in swaps for swap in board_swaps]
+
+
+def rating_variance_improvement(league_mean, n_boards, swap):
+    def score(a, b):
+        return variance(league_mean, [a, b])
+
+    pa, pb = swap
+    a_mean = pa.team.getMean()
+    b_mean = pb.team.getMean()
+    initial_score = score(a_mean, b_mean)
+
+    # calculating change in mean if we swapped the players.
+    rating_diff = pb.rating - pa.rating
+    a_mean = a_mean + rating_diff/n_boards
+    b_mean = b_mean - rating_diff/n_boards
+    new_score = score(a_mean, b_mean)
+
+    # lower is better
+    return new_score - initial_score
+
+
+def get_best_swap(swaps, fun):
+    best_swap = min(swaps, key=fun)
+    return best_swap, fun(best_swap)
+
+
+def perform_swap(swap):
+    pa, pb = swap
+    ta = pa.team
+    tb = pb.team
+    board = pa.board
+    ta.changeBoard(board, pb)
+    tb.changeBoard(board, pa)
+
+
+def update_swaps(swaps, swap_performed, teams):
+    pa, pb = swap_performed
+    affected_players = pa.team.boards + pb.team.boards
+    # remove all swaps involving players affected by the swap.
+    swaps = [swap for swap in swaps
+             if not (swap[0] in affected_players or swap[1] in affected_players)]
+
+    # find new neutral swaps involving the players affected by swap.
+    for player in affected_players:
+        board = player.board
+        players_on_board = [team.boards[board] for team in teams
+                            if not team.boards[board] in affected_players]
+        swaps.extend([(player, p) for p in players_on_board
+                      if is_neutral_swap((player, p))])
+
+    swaps.extend([swap for swap in zip(pa.team.boards, pb.team.boards)
+                  if is_neutral_swap(swap)])
+
+    return swaps
+
+
+def reduce_variance(teams):
+    # players = sum([team.boards for team in teams], [])
+
+    league_mean = sum([team.getMean() for team in teams]) / len(teams)
+    n_boards = len(teams[0].boards)
+
+    swaps = get_swaps(teams)
+
+    eval_fun = partial(rating_variance_improvement, league_mean, n_boards)
+    best_swap, swap_value = get_best_swap(swaps, eval_fun)
+
+    # infinite loop possibility here?
```

Comment:
I will remove that comment.  It's earlier when it really was getting into an infinite loop.  I added `epsilon` and `max_iterations` to prevent the infinite loops.  The problem was with floating point arithmetic.  It was getting to 1e-12 or so but never 0.  

I'm actually not even sure how that's happening but but didn't really look into it.  I would have thought that after a certain number of iterations all swaps should be negative not just less than `epsilon`.  But it ends up getting stuck in a state where there's one swap that's like 1e-12 both ways!